src/HTTPure/Body.purs

```diff
@@ -1,46 +1,91 @@
 module HTTPure.Body
   ( class Body
+  , RequestBody
   , defaultHeaders
   , write
   , read
-  , toString
   , toBuffer
+  , toStream
+  , toString
   ) where
 
 import Prelude
 import Data.Either (Either(Right))
+import Data.Maybe (Maybe(Just, Nothing))
 import Effect (Effect)
-import Effect.Class (liftEffect)
 import Effect.Aff (Aff, makeAff, nonCanceler)
-import Effect.Ref (read) as Ref
-import Effect.Ref (new, modify)
+import Effect.Class (liftEffect)
+import Effect.Ref (Ref)
+import Effect.Ref (read, modify, new, write) as Ref
 import HTTPure.Headers (Headers, header)
-import Node.Buffer (toString) as Buffer
 import Node.Buffer (Buffer, concat, fromString, size)
+import Node.Buffer (toString) as Buffer
 import Node.Encoding (Encoding(UTF8))
 import Node.HTTP (Request, Response, requestAsStream, responseAsStream)
-import Node.Stream (write) as Stream
 import Node.Stream (Stream, Readable, onData, onEnd, writeString, pipe, end)
+import Node.Stream (write) as Stream
 import Type.Equality (class TypeEquals, to)
 
+type RequestBody =
+  { buffer :: Ref (Maybe Buffer)
+  , stream :: Readable ()
+  , string :: Ref (Maybe String)
+  }
+
 -- | Read the body `Readable` stream out of the incoming request
 read :: Request -> Readable ()
 read = requestAsStream
```

Comment:
I think should be now updated to create the refs and return an `Effect RequestBody` instead, i.e.:

```purs
read :: Request -> Effect RequestBody
read = do
    buffer <- Ref.new Nothing
    string <- Ref.new Nothing
    pure { buffer, string, stream: Body.read request }
```

Reply:
Good idea. Updated.